expression/builtin_cast.go

```diff
@@ -1934,8 +1934,8 @@ func WrapWithCastAsDecimal(ctx sessionctx.Context, expr Expression) Expression {
 		return expr
 	}
 	tp := types.NewFieldType(mysql.TypeNewDecimal)
-	tp.SetFlen(expr.GetType().GetFlen())
-	tp.SetDecimal(expr.GetType().GetDecimal())
+	tp.SetFlenUnderLimit(expr.GetType().GetFlen())
+	tp.SetDecimalUnderLimit(expr.GetType().GetDecimal())
```

Comment:
Why not use `UpdateFlenAndDecimalUnderLimit` here?

Reply:
besides, UpdateFlenAndDecimalUnderLimit aims to maintain valid when update the flen or decimal, where the flen and decimal should be not be -1.